src/memory.jl

```diff
@@ -12,14 +12,15 @@ using Base.Threads: SpinLock
 const memory_lock = SpinLock()
 
 # the above spinlocks are taken around code that might gc, which might cause a deadlock
-# if we try to acquire from the finalizer too. avoid that by temporarily disabling the GC.
-macro without_gc(ex)
+# if we try to acquire from the finalizer too. avoid that by temporarily disabling finalizers.
```

Comment:
```suggestion
# if we try to acquire from the finalizer too. avoid that by temporarily disabling running finalizers,
# concurrently on this thread.
```

Chatting with Jameson about this, `jl_gc_enable_finalizer` will only inhibit running finalizers on the calling thread. They can still run on other threads (which should be fine). 

That is also the reason why the interleaving I was worried about can't happen.